src/utils/dispatcher.ts

```diff
@@ -57,17 +96,43 @@ const dispatcher: Dispatcher = {
     getServerSnapshot ? getServerSnapshot() : getSnapshot(),
 
   [__REACT_PREPARE__]: {
+    errorHandler,
     context: {},
+    hookPromises: [],
+    syncHookPromises: [],
+    preparedHookIdentifiers: [],
   },
-};
+});
 
-export const setDispatcherContext = (context: PrepareContext): void => {
+export const setDispatcherContext = (
+  dispatcher: Dispatcher,
+  context: PrepareContext,
+): void => {
   dispatcher[__REACT_PREPARE__].context = context;
 };
 
-export const registerDispatcher = (): void => {
+export const registerDispatcher = (dispatcher: Dispatcher): void => {
   ReactInternals.ReactCurrentDispatcher.current = dispatcher;
 };
 
-export const dispatcherIsRegistered = (): boolean =>
-  ReactInternals.ReactCurrentDispatcher.current === dispatcher;
+export const popPreparedHookPromises = (
+  dispatcher: Dispatcher,
+): Promise<unknown>[] => {
+  const promises = dispatcher[__REACT_PREPARE__].hookPromises;
+  dispatcher[__REACT_PREPARE__].hookPromises = [];
+  return promises;
```

Comment:
Yes, we basically add all `usePreparedEffect`-promises we find while parsing the component tree to the dispatcher. When the time comes to await the promises, we then pop all of them and await them at the same time using `Promise.all`. 

Normally this will happen at the end of `prepare`, after parsing the entire component-tree.

In the case of `usePreparedEffect`s with `runSync=true`, this will instead happen right away after parsing the component. The reason we need to pop an array of promises here is the case where one has multiple `runSync=true`-hooks in a single component, in which case we await these hooks simultainiously before parsing the children. (Ideally we would await the first one before parsing the second, but AFAIK this is not possible)